sdk/servicebus/Azure.Messaging.ServiceBus/samples/Sample15_MockingClientTypes.md

```diff
@@ -0,0 +1,619 @@
+// called.
+
+var mockQueueName = "MockQueueName";
+ServiceBusSender sender = client.CreateSender(mockQueueName);
+
+foreach (ServiceBusMessage message in testSessionMessages)
+{
+    await sender.SendMessageAsync(message);
+}
+
+TimeSpan maxWait = TimeSpan.FromSeconds(30);
+
+ServiceBusSessionReceiver sessionReceiver = await client.AcceptNextSessionAsync(mockQueueName, new ServiceBusSessionReceiverOptions(), CancellationToken.None);
+
+ServiceBusReceivedMessage receivedMessage = await sessionReceiver.ReceiveMessageAsync(maxWait, CancellationToken.None);
+
+BinaryData setState = new("MockState");
+await sessionReceiver.SetSessionStateAsync(setState, CancellationToken.None);
+BinaryData state = await sessionReceiver.GetSessionStateAsync(CancellationToken.None);
+
+Assert.AreEqual(setState, state);
+```
+
+## Sending and Receiving messages using topics and subscriptions
+
+```C# Snippet:ServiceBus_MockingTopicSubscriptionSend
+Mock<ServiceBusAdministrationClient> mockAdministrationClient = new();
+Mock<Response<TopicProperties>> mockTopicResponse = new();
+Mock<Response<SubscriptionProperties>> mockSuscriptionResponse = new();
+
+mockAdministrationClient
+    .Setup(client => client.CreateTopicAsync(
+        It.IsAny<CreateTopicOptions>(),
+        It.IsAny<CancellationToken>()))
+    .Callback<CreateTopicOptions, CancellationToken>((opts, ct) =>
+    {
+        TopicProperties mockTopicProperties = ServiceBusModelFactory.TopicProperties(opts);
+
+        mockTopicResponse.Setup(r => r.Value).Returns(mockTopicProperties);
+    })
+    .ReturnsAsync(mockTopicResponse.Object);
+
+mockAdministrationClient
+    .Setup(client => client.CreateSubscriptionAsync(
+        It.IsAny<CreateSubscriptionOptions>(),
+        It.IsAny<CancellationToken>()))
+    .Callback<CreateSubscriptionOptions, CancellationToken>((opts, ct) =>
+    {
+        SubscriptionProperties mockSubscriptionProperties = ServiceBusModelFactory.SubscriptionProperties(opts);
+
+        mockSuscriptionResponse.Setup(r => r.Value).Returns(mockSubscriptionProperties);
+    })
+    .ReturnsAsync(mockSuscriptionResponse.Object);
+
+var adminClient = mockAdministrationClient.Object;
+
+Mock<ServiceBusClient> mockServiceBusClient = new();
+Mock<ServiceBusSender> mockSender = new();
+Mock<ServiceBusReceiver> mockReceiver = new();
+
+mockServiceBusClient
+    .Setup(client => client.CreateSender(It.IsAny<string>()))
+    .Returns(mockSender.Object);
+
+mockSender
+    .Setup(sender => sender.SendMessageAsync(
+        It.IsAny<ServiceBusMessage>(),
+        It.IsAny<CancellationToken>()))
+    .Returns(Task.CompletedTask);
+
+mockServiceBusClient
+    .Setup(client => client.CreateReceiver(
+        It.IsAny<string>(),
+        It.IsAny<string>()))
+    .Returns(mockReceiver.Object);
+
+async IAsyncEnumerable<ServiceBusReceivedMessage> mockReturn()
+{
+    ServiceBusReceivedMessage message = ServiceBusModelFactory.ServiceBusReceivedMessage(
```

Comment:
We probably want to have a list of messages that we `foreach` over and return.   I can see this being a disconnect for customers without depth of knowledge.   Custom iteration isn't a common thing for a lot of less advanced scenarios.